### ISSUE TITLE
[BUG] Release 3.4 doesn`t work when running it as JAR file

### ISSUE THREAD

**Issue by alex231330** [odpi/egeria#6020]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

While building the new Egeria 3.4 Release with the following command line:

`mvn -B -U -Dmaven.test.skip -Drat.skip=true -Dmaven.javadoc.skip=true -DuseMavenCentral package`

As a result of build process i get a JAR, which i try to run, it crashed with ModuleNotFound. I`ve done some research and found out that there is some issue with local build process. Maybe maven forget to compile smth, idk. 

But if i will run Egeria as java application (without building jar) with that command line below:

`{javaPath}\java.exe -XX:TieredStopAtLevel=1 -noverify -Dspring.output.ansi.enabled=always "-javaagent:C:\Program Files\JetBrains\IntelliJ IDEA 2021.2.2\lib\idea_rt.jar=63544:C:\Program Files\JetBrains\IntelliJ IDEA 2021.2.2\bin" -Dcom.sun.management.jmxremote -Dspring.jmx.enabled=true -Dspring.liveBeansView.mbeanDomain -Dspring.application.admin.enabled=true -Dfile.encoding=UTF-8 @C:\Users\{user}\AppData\Local\Temp\idea_arg_file907069212 org.odpi.openmetadata.serverchassis.springboot.OMAGServerPlatform`

Everything works fine. 

### Expected Behavior

I expect that both way to run Egeria platform will work the same way

### Steps To Reproduce

Described in the current behavior

### Environment

```markdown
- Egeria: 3.4 Release
- OS: Windows 10
- Java: liberica 11.0.13
- Additional connectors and integration: CruxDB Egeria connector
```

**Comment by mandy-chessell** [comment]
Just checking that this is a change in release 3.4 and release 3.3 worked ok?

**Comment by planetf1** [comment]
A few thoughts
* The mvn invocation is skipping the execution of tests. We've seen problems in this area as with #6008 #6946 - it's possible there are other issues as it's not a scenario we test. 
* There was a change to the way the janusgraph connector is built/packaged
* We do not build or test on Windows 10. We do now recommend use of WSL2 (ie a full linux environment within w10) but do not test it regularly/for each release

So we need to :
* Check if the issue occurs without the additional maven flags (ie just use `mvn clean install`)
* Check if it's a regression in 3.4
* Check if it's specific to windows 10
* See more details on the failure you get - those additional flags are quite spring specific suggestion you see an issue with some of it's standard behaviour?

**Comment by planetf1** [comment]
I'll try out your mvn invocation on 3.4 also to see if I get any problems

**Comment by alex231330** [comment]
Well, we will try to build this under another OS enviroment. 
> Just checking that this is a change in release 3.4 and release 3.3 worked ok?
In the 3.3 release there was no such problem

**Comment by alex231330** [comment]
> * Check if the issue occurs without the additional maven flags (ie just use `mvn clean install`)

With thouse flags the problems still occurs

**Comment by planetf1** [comment]
I just tried those flags on macOS (my dev machine) and it worked fine.
Can you clarify what environment you were using under Windows - was it WSL2? Which linux distro?  Or we you building natively?

**Comment by planetf1** [comment]
From your build flags one area to check is spring changes. Here is 3.4 vs 3.3
```
jonesn:egeria/ (egeria-release-3.4) $ cat pom.xml | grep '<spring'                                       [10:40:41]
        <spring-boot.version>2.5.6</spring-boot.version>
        <spring.ldap.core>2.3.4.RELEASE</spring.ldap.core>
        <spring-security.version>5.5.3</spring-security.version>
        <spring-data.version>2.5.6</spring-data.version>
        <springdoc.version>1.5.12</springdoc.version>
        <spring.version>5.3.12</spring.version>
        <springdoc-plugin.version>1.3</springdoc-plugin.version>
jonesn:egeria/ (egeria-release-3.4) $ git checkout egeria-release-3.3                                    [10:40:55]
Updating files: 100% (1530/1530), done.
Switched to branch 'egeria-release-3.3'
jonesn:egeria/ (egeria-release-3.3) $ cat pom.xml | grep '<spring'                                       [10:41:03]
        <spring-boot.version>2.5.5</spring-boot.version>
        <spring.ldap.core>2.3.4.RELEASE</spring.ldap.core>
        <spring-data.version>2.5.5</spring-data.version>
        <spring-security.version>5.5.2</spring-security.version>
        <springdoc.version>1.5.12</springdoc.version>
        <spring.version>5.3.10</spring.version>
        <springdoc-plugin.version>1.3</springdoc-plugin.version>
```

**Comment by alex231330** [comment]
There is a little bit difference problem - can you try to create several entites and a connection between them?

**Comment by alex231330** [comment]
About my spring verisions:

1. <spring-boot.version>2.5.6</spring-boot.version>
2. <spring.ldap.core>2.3.4.RELEASE</spring.ldap.core>
3. <spring-security.version>5.5.3</spring-security.version>
4. <spring-data.version>2.5.6</spring-data.version>
5. <springdoc.version>1.5.12</springdoc.version>
6. <spring.version>5.3.12</spring.version>
7. <springdoc-plugin.version>1.3</springdoc-plugin.version>

**Comment by planetf1** [comment]
I was able to run the jar on WSL2 also

**Comment by planetf1** [comment]
> There is a little bit difference problem - can you try to create several entites and a connection between them?

Can you clarify?
Without a log I was assuming you mean the jar crashes on startup
Are you saying it crashes when you do a particular operation?
If so can you share a script to reproduce? 
Can you share the errors/exception you see?
And also platform - is it WSL2 on windows or native?

**Comment by planetf1** [comment]
> About my spring verisions:

Yes - that is what we use for 3.4, we both have the same
I posted a comparison with the 3.3 release since you mentioned it worked there, and the parameters you used to get it working were quite spring-framework specific, so I was wondering if the spring versions might be a factor

**Comment by alex231330** [comment]
Well, please give me some time - i will create a java script to reproduce it

**Comment by alex231330** [comment]
So, there is another, faster way - could you please send us your build jar file? We use our custom backend to work with egeria. 

**UPDATE:** 
Build on linux os (native) stikk occurs that error

**Comment by planetf1** [comment]
I've just built on Windows 10 21H2 natively using the latest Adoptium (Temurin) JDK 17 - this is the jdk distro we mostly use including for formal builds - though they are usually at 11 . Again I used your maven command

The build took about twice (28 min) as long natively vs WSL2 (12 min) but succeeded ok (there are some test issues with Windows, which is why we recommend WSL2 - contributions are welcome to improve those tests, setup automated windows testing etc...)

Also here the jar starts up fine

**Comment by planetf1** [comment]
Our release process attaches the build output as an attachment

The run we used for release 3.4 is here -> https://github.com/odpi/egeria/actions/runs/1515414932

**Comment by planetf1** [comment]
Can you clarify what you are seeing. In all these environments the jar runs ok.
It could be jdk difference.. or it's that you get an error after the initial startup when you are processing requests
Can you send the message you get on stdout/stderr relating to the missing module?

**Comment by alex231330** [comment]
So, i have an another error report update:

When i add the install flag to maven, i got test failures
[ORGODP~1.TXT](https://github.com/odpi/egeria/files/7649639/ORGODP.1.TXT)

Also i read build log file line-by-line an found this:

`Failure to find org.apache.felix:maven-bundle-plugin:pom:2.4.0-build002 in https://oss.sonatype.org/content/groups/public was cached in the local repository, resolution will not be reattempted until the update interval of sonatype-nexus-archon has elapsed or updates are forced
`
But this message is just warning and don`t break the whole building process

And I collect the log about error, which happens both in your jar file and my jar file. This problem is JDK independent
[asset connection create error.txt](https://github.com/odpi/egeria/files/7649667/asset.connection.create.error.txt)

**Comment by planetf1** [comment]
Thanks
 * The test failure looks to be WIndows specific. Currently we're not testing/supporting windows. We are instead suggesting the use of WSL2 under Windows. If you would like to raise the windows issue, and work through the fix, support windows, and ensure all the tests work on Windows then that would be a valuable contribution. As part of that we'd need to ensure we do sufficient tests within the PR process to avoid future regressions. We haven't had the strong requirement or resource to work on windows native so far.. and also noted many applications/projects are now directing users to WSL2

**Comment by planetf1** [comment]
I can now see the error as:
```
ri Dec 03 15:44:36 MSK 2021 paukd-platform-local Exception OMRS-AUDIT-9016 The topic listener for the Governance Engine OMAS service caught an unexpected exception java.lang.NoSuchMethodError with message 'boolean org.odpi.openmetadata.repositoryservices.connectors.stores.metadatacollectionstore.properties.typedefs.RelationshipDef.getMultiLink()'
Fri Dec 03 15:44:36 MSK 2021 paukd-platform-local Exception OMRS-AUDIT-9016 Supplementary information: log record id 02111461-df50-46bc-96ea-14b57c917d21 java.lang.NoSuchMethodError returned message of 'boolean org.odpi.openmetadata.repositoryservices.connectors.stores.metadatacollectionstore.properties.typedefs.RelationshipDef.getMultiLink()' and stacktrace of
java.lang.NoSuchMethodError: 'boolean org.odpi.openmetadata.repositoryservices.connectors.stores.metadatacollectionstore.properties.typedefs.RelationshipDef.getMultiLink()'
        at org.odpi.openmetadata.commonservices.repositoryhandler.RepositoryHandler$RelationshipAccumulator.filterOutDuplicateRelationships(RepositoryHandler.java:262)
        at org.odpi.openmetadata.commonservices.repositoryhandler.RepositoryHandler$RelationshipAccumulator.getRelationships(RepositoryHandler.java:223)
        at org.odpi.openmetadata.commonservices.repositoryhandler.RepositoryHandler.getRelationshipsByType(RepositoryHandler.java:4685)
        at org.odpi.openmetadata.commonservices.repositoryhandler.RepositoryHandler.getUniqueRelationshipByType(RepositoryHandler.java:5703)
```

**Comment by planetf1** [comment]
This looks as if is a method api change.
Is this happening at runtime with the crux connector

I can see getMultiLink  was added in https://github.com/odpi/egeria/commit/d15acd1f26b5bc241724bf38bfcb3954c4fa8ab2#diff-e91dc843cf86cc080bf61fc498ae66f73cccfbd7a78a2ff7a27d0366d37b4f44

I wonder if this specific to the xtdb connector environment and resulting in your resolving to an older version of the class
@cmgrote @mandy-chessell

**Comment by planetf1** [comment]
Another observation - the exception is occuring from Governance Engine OMAS. Do you need to run this OMAS? If not you could disable it? And I suspect you don't see other issues (though it's still a bug)

**Comment by planetf1** [comment]
@mandy-chessell As this is in the area of duplicate handling at the relationship modifications can you take a look?

**Comment by mandy-chessell** [comment]
Looks like it is using an OMRS from a different release .... This method
`boolean org.odpi.openmetadata.repositoryservices.connectors.stores.metadatacollectionstore.properties.typedefs.RelationshipDef.getMultiLink()`
 was added 29th October 2021

**Comment by planetf1** [comment]
@alex231330 I'm also unsure why it does not find the class.

Do you get the issue when using other repo like JanusGraph, or is it just when using xtdb (I'd assume both - but worth checking as currently xtdb is built against 3.3 )
Can you share the script/curl commands to reproduce the issue?

**Comment by planetf1** [comment]
@alex231330 Can you check your classpath when you run the jar, ie anywhere pointed to by spring's 'loader.path' ? By default this is the server/lib directory. Are there any v3.3 classes there, specifically its 'org.odpi.egeria:repository-services-apis'  containing 'repository-services-apis-3.4.jar' where this class is defined. do you for example have any other connectors that have built with dependencies and are including version 3.3 jars?

**Comment by mandy-chessell** [comment]
@alex231330 For example, is there a home-grown connector - such as the security connector - in the LIBPATH that is linking an older version of the repository-services-apis module?

**Comment by alex231330** [comment]
> @alex231330 For example, is there a home-grown connector - such as the security connector - in the LIBPATH that is linking an older version of the repository-services-apis module?

Yes, we have a crux db connector which we load via -Dloaderpath

**Comment by cmgrote** [comment]
Could the problem here be that the XT connector is using the repository-services-api as "provided" without being recompiled against the newer release?

I would have expected that not to be an issue, because if it is it will mean every connector must be recompiled against the release to work (which sounds a bit more like a breaking change that we've included in a minor release)?

**Comment by cmgrote** [comment]
Here's a pre-built version of the XT connector jar that is now built against the 3.4 Egeria release -- might be worth testing using this connector jar with your startup sequence just to confirm whether the problem still exists, or if it is resolved (suggesting that indeed connectors must be recompiled against the release even if just specifying dependencies as `provided`...)

https://oss.sonatype.org/content/repositories/snapshots/org/odpi/egeria/egeria-connector-xtdb/3.4-SNAPSHOT/egeria-connector-xtdb-3.4-20211206.101230-6-jar-with-dependencies.jar

Based on what you confirm, I can then decide whether to release the 3.4 version of the XT connector as-is to address this (rather than waiting on further details on the other exception that we mentioned a couple weeks back on a dev call -- though perhaps the two are related?)

**Comment by planetf1** [comment]
I presume the issue here could indeed be that @cmgrote - whilst the implementation of a method changing would be handled well with provided, it does not address method signature changes

**Comment by planetf1** [comment]
From the exception / stacktrace I'm struggling to see it as a connector build issue using provided ... but it's definitely worth trying. Also an older connector that has the library included would be an issue

**Comment by alex231330** [comment]
> Another observation - the exception is occuring from Governance Engine OMAS. Do you need to run this OMAS? If not you could disable it? And I suspect you don't see other issues (though it's still a bug)

Hello again! Thanks for wainig. We resolve that issue by disabling OMAS that you mentioned - Gouvernance Engine. Everything works OK without it. But i think that we need to open another issue about it

**Comment by planetf1** [comment]
Glad to hear you are working now...  however were you able to find any reference to an older jar around (either an earlier 3.4-SNAPSHOT or more likely a 3.3 or earlier version). I've not yet figured out a particular mechanism that would explain the error you saw
